### PR TITLE
justin lite

### PR DESCRIPTION
**Pull request by jadrubio**
### Description

-

### Related Jira Tickets

-

### Comments 

-

Please ensure all items are complete before requesting a review:

- [ ] **Code Completed**: All code for the feature or bugfix has been implemented.
- [ ] **Unit Tests Added**: Unit tests are written for all new functionality, covering edge cases where applicable.
- [ ] **Integration Tests (where necessary)**: Integration tests are added for interactions between modules or components.
- [ ] **All Tests Passing**: All unit and integration tests are passing.
- [ ] **Type Safety Checked**: Ensure TypeScript types are correctly defined and used. Avoid `any` types without proper justification.
- [ ] **Typedoc Comments Added**: Ensure all functions, classes, interfaces, and types are documented using TypeDoc syntax.
- [ ] **Code Formatting**: All code files are formatted according to the project’s standards (e.g., Prettier, ESLint) and no linting errors remain.
- [ ] **No Console Logs**: Remove any `console.log` or other debugging output from the code.
- [ ] **Documentation Updated**: Update relevant project documentation (e.g., README, configuration guides) if necessary.

---

### CODE REVIEW

File: src/JustInLite.ts [hunk 1]
```diff
@@ -1,98 +1,264 @@
...
+      }
+
+      if (next.has(uniqueIdentifier)) {
+        const msg = `addUsers: duplicate uniqueIdentifier "${uniqueIdentifier}" (again at index ${i}).`;
+        Log.error(msg);
+        throw new Error(msg);
+      }
+
+      const attrs =
+        'attributes' in anyItem ? (anyItem.attributes ?? {}) :
```
Comment (numerator) [inline src/JustInLite.ts:128]: I see why you were asking about this. I'm not wedded to forcing NewUserRecords to using `initialAttributes` as the key name. If we're going to allow both I might actually prefer to just have one and have it be `attributes`.

File: src/JustInLite.ts [hunk 2]
```diff
@@ -1,98 +1,264 @@
...
-  public async cleanseDB(): Promise<void> {
-    await DataManager.getInstance().clearCollection(EVENT_QUEUE);
-    await DataManager.getInstance().clearCollection(USERS);
-    Log.info('DB cleansed');
+  public async registerEventHandlers(
+    eventType: string,
+    handlers: string[],
+  ): Promise<void> {
+    if (this.eventDefinitions.has(eventType)) {
+      throw new Error(`Event "${eventType}" already registered; pass overwriteExisting=true to replace.`);
```
Comment (numerator) [inline src/JustInLite.ts]: where would they pass this? It doesn't seem to be an argument for this function or for the constructor/getInstance

File: src/JustInLite.ts [hunk 3]
```diff
@@ -1,98 +1,264 @@
...
+  // ────────────────────────────────────────────────────────────────────────────
+
   /**
-   * Resets all singleton instances used by the framework.
-   *
-   * This should be called at the start of a stateless execution context,
-   * such as in Google Cloud Functions or tests.
+   * Adds users for the current invocation (serverless-safe).
+   * Accepts either `JUser[]` or `NewUserRecord[]`.
+   * Replaces the in-memory set each call (atomic), requires `uniqueIdentifier`,
```
Comment (numerator) [inline src/JustInLite.ts:97]: Since this replaces the entire set of users, should the function be called "loadUsers" or something like that? This would apply to JustInWrapper too--whatever we decide should be the same for both. And yes that's a breaking change. But if nobody's using it does it really break anything? Tree in the woods, etc.

File: src/__tests__/JustInLite.integration.test.ts [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    hasHandlersForEventTypeStub.restore();
+    executeEventForUsersStub.restore();
+    logInfoStub.restore();
+    logWarnStub.restore();
+    logErrorStub.restore();
+    logDevStub.restore();
+  });
+
+  describe('addUsers', () => {
+    it('adds JUser[] and replaces previous set', async () => {
```
Comment (numerator) [inline src/__tests__/JustInLite.integration.test.ts]: Ah. So you can add either JUsers or NewUserRecords? Not sure why we support both.

File: src/__tests__/JustInLite.integration.test.ts [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    });
+
+    it('throws on missing uniqueIdentifier', async () => {
+      await expect(
+        lite.addUsers([{ id: 'no-uid', attributes: {} } as unknown as JUser])
+      ).rejects.toThrow("missing required 'uniqueIdentifier'");
+      expect(logErrorStub.calledWithMatch("missing required 'uniqueIdentifier'")).toBe(true);
+    });
+
+    it('throws on duplicate uniqueIdentifier in same call', async () => {
```
Comment (numerator) [inline src/__tests__/JustInLite.integration.test.ts:98]: do we have/need a test for not throwing on sequential calls?

File: src/__tests__/JustInLite.integration.test.ts [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+          { id: 'x', uniqueIdentifier: 'dup', attributes: {} } as JUser,
+          { id: 'y', uniqueIdentifier: 'dup', attributes: {} } as JUser,
+        ])
+      ).rejects.toThrow('duplicate uniqueIdentifier "dup"');
+      expect(logErrorStub.calledWithMatch('duplicate uniqueIdentifier "dup"')).toBe(true);
+    });
+  });
+
+  describe('registration', () => {
+    it('registerEventHandlers caches and forwards to manager', async () => {
```
Comment (numerator) [inline src/__tests__/JustInLite.integration.test.ts]: I don't understand what this is doing or why caching is involved.

Comment (numerator) [reply src/__tests__/JustInLite.integration.test.ts]: O I think I see. `registerEventHandlersStub` is stubbing the eventmanager function, so you're seeing that that stub was called and also that lite is maintaining it's own list. Not sure why we have both, but OK.

File: src/__tests__/JustInLite.integration.test.ts [hunk 7]
```diff
@@ -0,0 +1,188 @@
+import sinon from 'sinon';
```
Comment (numerator) [inline src/__tests__/JustInLite.integration.test.ts:1]: Would it be a good practice to name or state in comments what is integrating with what for integration tests? Like this one is testing integration of JustInLite with EventQueue (?) but not with EventHandlerManager or EventExecutor (?). So the file name might be JustInLite.EventQueue.integration.tests.ts? Just a thought.

File: src/event/__tests__/event-executor.unit.test.ts [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+      await executeEventForUsers(event, users, handlerManager);
+
+      expect(beforeStub.calledOnceWith(event)).toBe(true);
+      expect(executeTaskStub.calledTwice).toBe(true);
+      expect(executeTaskStub.firstCall.calledWith(mockTask, event, users[0])).toBe(true);
+      expect(executeTaskStub.secondCall.calledWith(mockTask, event, users[1])).toBe(true);
+      expect(afterStub.calledOnceWith(event)).toBe(true);
+    });
+
+    it('logs beforeExecution error and still runs per-user and afterExecution', async () => {
```
Comment (numerator) [inline src/event/__tests__/event-executor.unit.test.ts:100]: Raises an interesting question--should the 3PD be able to override this default behavior--i.e., have something in BE that aborts the rest of the event. I think it only makes sense for BE. Or is there already a way to do that, just not tested here?

File: src/event/__tests__/event-executor.unit.test.ts [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+
+      await executeEventForUsers(event, users, handlerManager);
+
+      expect(executeTaskStub.calledTwice).toBe(true);
+      expect(logErrorStub.calledWith(
+        'Execution error for "taskA" on user "u1" (event "TEST_EVENT"): Error: u1-fail'
+      )).toBe(true);
+    });
+
+    it('logs afterExecution error and completes', async () => {
```
Comment (numerator) [inline src/event/__tests__/event-executor.unit.test.ts:144]: Another interesting question. If actions are completed in afterExecution (e.g., batch send of notifications) and those actions failed, should they affect the results of the affected user? Perhaps not in the sense of rewriting those results, but in terms of providing a paper trail for an analyst to reconstruct the failed action? I think right now AE logs are not written to results, but perhaps they should be, and each event execution should have an identifier (publishedTimestamp might be good enough for this) that would allow for this kind of reconstruction.

File: src/event/__tests__/event-queue.unit.test.ts [hunk 10]
```diff
@@ -470,20 +446,16 @@ describe('Event Queue', () => {
...
       getAllInCollectionStub.onSecondCall().resolves([]);
 
       await EventQueue.processEventQueue();
 
-      expect(logDevStub.calledWith('"beforeExecution" not found for handler "task1".')).toBe(true);
-      expect(logDevStub.calledWith('"afterExecution" not found for handler "task1".')).toBe(true);
+      expect(executeEventForUsersStub.calledOnce).toBe(true);
     });
 
     it('should handle lifecycle method execution errors', async () => {
```
Comment (numerator) [inline src/event/__tests__/event-queue.unit.test.ts:462]: Is this test label still accurate? I don't see any errors being thrown or caught in this test.

File: src/event/__tests__/event-queue.unit.test.ts [hunk 11]
```diff
@@ -529,6 +493,7 @@ describe('Event Queue', () => {
       getAllInCollectionStub.onFirstCall().resolves([mockEvent]);
       getAllInCollectionStub.onSecondCall().resolves([]);
       getHandlersForEventTypeStub.returns([]);
+      executeEventForUsersStub.resolves();
```
Comment (numerator) [inline src/event/__tests__/event-queue.unit.test.ts:497]: Is this test label correct? Seems like it should be 'should archive event successfully if archiving enabled' (and it's disabled by lite, yes?

Comment (numerator) [reply src/event/__tests__/event-queue.unit.test.ts:497]: Or I suppose what this is actually testing is that the event is removed from the queue after processing, but not that it was actually archived.

File: src/event/event-executor.ts [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+  const beforeRan = new Set<string>();
+  const afterRan = new Set<string>();
+
+  for (const handlerName of handlerNames) {
+    const task = getTaskByName(handlerName);
+    const rule = getDecisionRuleByName(handlerName);
+
+    // BEFORE (once per handler)
+    if (!beforeRan.has(handlerName)) {
```
Comment (numerator) [inline src/event/event-executor.ts:35]: Not sure why this is needed. Idempotency?

File: src/event/event-queue.ts [hunk 13]
```diff
@@ -83,23 +84,7 @@ export const processEventQueue = async (): Promise<void> => {
...
-          }
-        }
-
-        for (const handler of eventHandlerManager.getHandlersForEventType(event.eventType)) {
-          await processExecutionLifecycle(handler, event, "afterExecution")
-        }
+        await executeEventForUsers(event, users, eventHandlerManager);
 
         try {
           await archiveEvent(event);
```
Comment (numerator) [inline src/event/event-queue.ts:90]: Does this get called in lite? And we just let it log an error?

File: src/event/event-queue.ts [hunk 14]
```diff
@@ -83,23 +84,7 @@ export const processEventQueue = async (): Promise<void> => {
...
-            await processHandlers(event, user);
-          } catch (error) {
-            Log.error(
-              `Error processing event "${event.eventType}" with ID: ${event.id} for user ${user.id}: ${error}`
-            );
-          }
-        }
-
-        for (const handler of eventHandlerManager.getHandlersForEventType(event.eventType)) {
-          await processExecutionLifecycle(handler, event, "afterExecution")
```
Comment (numerator) [inline src/event/event-queue.ts:101]: I see that this and similar functions are still present, but should they be deleted? They are obsolete no?

**Review by numerator — COMMENTED** [review]
Most of these comments are really just discussion topics or questions, and not requests for changes. Would be good to triage and decide what might need tweaking and what to kick down the road (and put on a list)

File: src/JustInLite.ts [hunk 15]
```diff
@@ -1,98 +1,264 @@
...
+   * and throws on duplicates. Returns the normalized `JUser[]`.
    */
-  public static reset(): void {
-    JustInWrapper['killInstance']?.();
-    DataManager['killInstance']?.();
-    ChangeListenerManager['killInstance']?.();
-    JustinLiteWrapper.instance = null;
+  public async addUsers(users: JUser[] | NewUserRecord[]): Promise<JUser[]> {
```
Comment (peiyaoh) [inline src/JustInLite.ts]: I need a refresher on why we want to support JUser? I believe you mentioned it but I forgot.

File: src/JustInLite.ts (hunk 2 above)
Comment (peiyaoh) [reply src/JustInLite.ts]: I still have the same question. Where should a dev go/look at to change if they run into this error message?

File: src/JustInLite.ts [hunk 16]
```diff
@@ -1,98 +1,264 @@
...
-        throw new Error('Timeout while waiting for event queue to empty ' + timeout);
+  public async publishEvent(
+    eventType: string,
+    generatedTimestamp: Date,
+    eventDetails?: object,
+    idempotencyKey?: string
+  ): Promise<void> {
+    // Optional in-memory idempotency for cloud runs
+    if (idempotencyKey) {
+      if (this.processedKeys.has(idempotencyKey)) {
```
Comment (peiyaoh) [inline src/JustInLite.ts:202]: Question: does this handle the case when first ping/event arrives -> handled, and then the second ping/event (for the same time) arrives? Or does it not happen this way ever?

It it correct to assume that the processedKeys are in-memory so will go away after the first ping? In that case, the second ping will still be processed?

File: src/__tests__/JustInLite.integration.test.ts [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+      expect(out.length).toBe(2);
+      expect(out[0]?.uniqueIdentifier).toBe('u1');
+      expect(out[0]?.attributes?.x).toBe(1);
+      expect(out[1]?.uniqueIdentifier).toBe('u2');
+    });
+
+    it('throws on missing uniqueIdentifier', async () => {
+      await expect(
+        lite.loadUsers([{ id: 'no-uid', attributes: {} } as unknown as JUser])
+      ).rejects.toThrow("missing required 'uniqueIdentifier'");
```
Comment (peiyaoh) [inline src/__tests__/JustInLite.integration.test.ts]: We should probably use a consistent message for both lite and also user-manager (addUser). I am fine using either version.

**Review by peiyaoh — DISMISSED** [review]
Left some comments.

**Review by numerator — APPROVED** [review]
